### ISSUE TITLE
Support binf.cljc library by adding ByteBuffer and Charset-related classes

### ISSUE THREAD

**Issue by sh54** [babashka/babashka#784]
**version**
v0.3.3

**platform**
macOS

**problem**
Fails when requiring in a dependency that does an `(:import (java.nio.charset Charset))`. 

Specifically the library https://github.com/helins/binf.cljc

**repro**
```clojure
(require '[babashka.deps :as deps])
(deps/add-deps '{:deps {io.helins/binf {:mvn/version "1.0.0-beta1"}}})
(require '[helins.binf :as binf])
```

**expected behavior**
Dependency can be required in successfully.

**abridged error from command line**
```
----- Error --------------------------------------------------------------------
Type:     java.lang.Exception
Message:  Unable to resolve classname: java.nio.charset.Charset
Location: helins/binf/string.cljc:12:1
----- Context ------------------------------------------------------------------
12: #?(:clj (:import (java.nio.charset Charset
    ^--- Unable to resolve classname: java.nio.charset.Charset
13:                                    StandardCharsets))))
```

**Comment by borkdude** [comment]
@sh54 I had to add a couple more classes. Before I merge to master, could you maybe explain why binf could be relevant to bb scripts/apps? Maybe @helins has something to add?

After adding the classes this works:

``` clojure
(require '[babashka.deps :as deps])
(deps/add-deps '{:deps {io.helins/binf {:mvn/version "1.0.0-beta1"}}})

(require '[helins.binf        :as binf]
         '[helins.binf.buffer :as binf.buffer])

(def my-buffer
  (binf.buffer/alloc 1024))
```

**Comment by helins** [comment]
From the `java.nio.charset` package are needed (but not available in BB): `Charset`, `StandardCharsets`, and `CoderResult`. Rest seems fine, I believe :)

I could see value in using BinF for you application/library, for instance something about interacting with endpoints that use a custom binary protocol and then using BB for writing lightweight CLI tooling. I would be interested in knowing more about real world use cases.

**Comment by borkdude** [comment]
I actually needed a couple more to make binf work: https://github.com/babashka/babashka/commit/60ec6ff8d4233f954833f4ef2d87de51de2fddfc

You can visit the commit and download a binary if you want to play with it.

**Comment by helins** [comment]
Indeed, I didn't notice all the `ByteBuffer` related classes you added were not available. However, you might want to try adding `java.nio.HeapByteBuffer` as well. Regular byte buffers delegates to that and something is not quite right.

Simply writing and reading an int:

```clojure
(require '[babashka.deps :as deps])
(deps/add-deps '{:deps {io.helins/binf {:mvn/version "1.1.0-beta0"}}})

(require '[helins.binf        :as binf]
         '[helins.binf.buffer :as binf.buffer])

(import 'java.nio.ByteBuffer)

(def view
     (binf/view (binf.buffer/alloc 1024)))

(binf/wa-b32 view
             0
             42)
             
(println (binf/ra-i32 view
                      0))
```

Error:

```shell
----- Error --------------------------------------------------------------------
Type:     clojure.lang.ExceptionInfo
Message:  Method putInt on class java.nio.HeapByteBuffer not allowed!
Location: /tmp/binf.clj:176:16

----- Context ------------------------------------------------------------------
                 ^--- Method putInt on class java.nio.HeapByteBuffer not allowed!
```

In contrast, using a `DirectByteBuffer` from `(ByteBuffer/allocateDirect)` works as it doesn't call any parent class I believe.

**Comment by helins** [comment]
A somewhat similar problem when trying to write strings:

```clojure
(binf/wr-string view
                "Hello")


(binf/ra-string view
                0
                (binf/position view))
```

Error:

```shell
----- Error --------------------------------------------------------------------
Type:     clojure.lang.ExceptionInfo
Message:  Method newEncoder on class sun.nio.cs.UTF_8 not allowed!
```

Weird, that charset is even interned in the `helins.binf.string` namespace, I don't understand why.

**Comment by helins** [comment]
Right, it might be because those are abstract methods. Does it makes sense?

**Comment by sh54** [comment]
> @sh54 I had to add a couple more classes. Before I merge to master, could you maybe explain why binf could be relevant to bb scripts/apps?

It is a neat library that can be used to read and write binary formats in both clojure and clojurescript. In my particular case I have my regular application that works with both clojure and clojurescript and contains code that deals with a basic binary format. I would like the possibility to run a few file related things from the cli via bb. I want a little bb script that loads in binf, requires in the namespace that deals with my binary format, then calls on the relevant functions to get things done. All of that has been coded but I am blocked on loading in binf. This is a nice way of exposing some functionality from my main application into a little helper program.

Aside from that I think a good case for babashka scripts is for doing stuff related to binary files. So far I have much preferred using binf to help with this rather than doing it all without libraries so it would be great if that library could be supported!

**Comment by borkdude** [comment]
@sh54 @helins  Could you test this binary: 

Linux: https://17223-201467090-gh.circle-artifacts.com/0/release/babashka-0.3.5-SNAPSHOT-linux-amd64.tar.gz
macOS: https://17222-201467090-gh.circle-artifacts.com/0/release/babashka-0.3.5-SNAPSHOT-macos-amd64.tar.gz

**Comment by borkdude** [comment]
Updated binaries:

https://17226-201467090-gh.circle-artifacts.com/0/release/babashka-0.3.5-SNAPSHOT-linux-amd64.tar.gz
https://17224-201467090-gh.circle-artifacts.com/0/release/babashka-0.3.5-SNAPSHOT-macos-amd64.tar.gz

I think these are good to go to run binf. Any last minute testing tomorrow would be nice.

**Comment by sh54** [comment]
The macOS snapshot works great! My little utility scripts run smoothly now.

A small side note, I had a path that used [aset-byte](https://clojuredocs.org/clojure.core/aset-byte) that bb bailed on. I was porting that code away so no real concern to me but its always nice to have access to as many of the core functions as possible. [Clojure source for them](https://github.com/clojure/clojure/blob/clojure-1.10.1/src/clj/clojure/core.clj#L3935-L3984). But that is a sci thing so I can file it there.

**Comment by borkdude** [comment]
@sh54 Added `aset-byte` and other variants in sci now, will be in the next bb release
